src/docker/shared.rs

```diff
@@ -488,6 +489,287 @@ impl Directories {
+    }
+
+    pub fn exit_static() {
+        // SAFETY: an atomic store.
+        unsafe {
+            CONTAINER.exit();
+        }
+    }
+
+    // when the `docker exec` command finished.
+    pub fn finish(&mut self, is_tty: bool, msg_info: &mut MessageInfo) {
+        // relax the no-timeout and lack of output
+        // ensure we have atomic ordering
+        if self.exists() {
+            let deleter = self.deleter.as_mut().unwrap();
+            if is_tty {
+                deleter.timeout = DEFAULT_TIMEOUT;
+            }
+            deleter.color_choice = msg_info.color_choice;
+            deleter.verbosity = msg_info.verbosity;
+        }
+
+        self.terminate();
+    }
+
+    pub fn finish_static(is_tty: bool, msg_info: &mut MessageInfo) {
+        // SAFETY: internally guarded by an atomic load.
+        unsafe {
+            CONTAINER.finish(is_tty, msg_info);
+        }
+    }
+
+    // terminate the container early. leaves the struct in a valid
+    // state, so it's async safe, but so the container will not
+    // be stopped again.
+    pub fn terminate(&mut self) {
+        if self.exists.swap(false, Ordering::SeqCst) {
+            let deleter = self.deleter.as_mut().unwrap();
+            let mut msg_info = MessageInfo::new(deleter.color_choice, deleter.verbosity);
+            container_stop(
+                &deleter.engine,
+                &deleter.name,
+                deleter.timeout,
+                &mut msg_info,
+            )
+            .ok();
+            container_rm(&deleter.engine, &deleter.name, &mut msg_info).ok();
+
+            self.deleter = None;
+        }
+    }
+}
+
+impl Drop for Container {
+    fn drop(&mut self) {
+        self.terminate();
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ContainerState {
+    Created,
+    Running,
+    Paused,
+    Restarting,
+    Dead,
+    Exited,
+    DoesNotExist,
+}
+
+impl ContainerState {
+    pub fn new(state: &str) -> Result<Self> {
+        match state {
+            "created" => Ok(ContainerState::Created),
+            "running" => Ok(ContainerState::Running),
+            "paused" => Ok(ContainerState::Paused),
+            "restarting" => Ok(ContainerState::Restarting),
+            "dead" => Ok(ContainerState::Dead),
+            "exited" => Ok(ContainerState::Exited),
+            "" => Ok(ContainerState::DoesNotExist),
+            _ => eyre::bail!("unknown container state: got {state}"),
+        }
+    }
+
+    #[must_use]
+    pub fn is_stopped(&self) -> bool {
+        matches!(self, Self::Exited | Self::DoesNotExist)
+    }
+
+    #[must_use]
+    pub fn exists(&self) -> bool {
+        !matches!(self, Self::DoesNotExist)
+    }
+}
+
+// the lack of MessageInfo is because we need a mutable reference to
+// store if we need to erase a line from stdout or stderr. since our
+// cleanup will not need to erase a line, we can just store the basic
+// MessageInfo configurations.
```

Comment:
```suggestion
// the lack of [MessageInfo] is because it'd require a mutable reference,
// since we don't need the functionality behind the [MessageInfo], we can just store the basic
// MessageInfo configurations.
```